### PR TITLE
Return expected wire order with qml.probs(), use same logic in QSV

### LINKED ISSUE #3741: [BUG] Confusing behavior with the `wires` argument of `qml.probs`
### Expected behavior

Consider a case where where a device's wires are ordered as `[2,0,1]` and we'd like to get the probabilities associated with these wires in the order `[0,1,2]`. That is, if a `qml.PauliX(wires=[0])` operation is applied, I would expect `qml.probs(wires=[0,1,2])` to return the probabilities `[0,0,0,0,1,0,0,0]` because `0` is the ordered as the most significant bit in the wire ordering.

In the two following code snippets, I expect the `[0,0,0,0,1,0,0,0]` probability vector to be returned when `qml.PauliX` is applied to wire `[0]`.

1. Explicitly specifying the operation's wire:

```python
dev = qml.device("default.qubit", wires=[2,0,1])
@qml.qnode(dev)
def circ():
    qml.PauliX(wires=[0])
    return qml.probs(wires=[0,1,2])

circ()
```

2. Specifying the operation's wire based on its relative position in `dev.wires`:

```python
dev = qml.device("default.qubit", wires=[2,0,1])
@qml.qnode(dev)
def circ():
    qml.PauliX(wires=dev.wires[1])
    return qml.probs(wires=[0,1,2])

circ()
```

I believe that this expectation should hold because I specify that the `qml.PauliX` operates upon the `[0]` wire and that `qml.probs(wires=[0,1,2])` should return the probabilities ordered with the `[0]` wire being the most significant bit.

## Case 1.

```python
dev = qml.device("default.qubit", wires=[2,0,1])
@qml.qnode(dev)
def circ():
    qml.PauliX(wires=[0])
    return qml.probs(wires=[0,1,2])

circ()
```

Returns:

```
tensor([0., 1., 0., 0., 0., 0., 0., 0.], requires_grad=True)
```

## Case 2:

```python
dev = qml.device("default.qubit", wires=[2,0,1])
@qml.qnode(dev)
def circ():
    qml.PauliX(wires=dev.wires[1])
    return qml.probs(wires=[0,1,2])

circ()
```

Returns:

```
tensor([0., 1., 0., 0., 0., 0., 0., 0.], requires_grad=True)
```

## Problem

In both cases, it appears that the `qml.PauliX` operation is applied to wire `[2]` instead of wire `[0]` like I intended.

### Additional information

_No response_

### Source code

_No response_

### Tracebacks

_No response_

### System information

```shell
Name: PennyLane
Version: 0.28.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /Users/brian/opt/anaconda3/lib/python3.9/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, requests, retworkx, scipy, semantic-version, toml
Required-by: PennyLane-Lightning, qNetVO

Platform info:           macOS-10.16-x86_64-i386-64bit
Python version:          3.9.13
Numpy version:           1.21.5
Scipy version:           1.9.1
Installed devices:
- lightning.qubit (PennyLane-Lightning-0.28.1)
- default.gaussian (PennyLane-0.28.0)
- default.mixed (PennyLane-0.28.0)
- default.qubit (PennyLane-0.28.0)
- default.qubit.autograd (PennyLane-0.28.0)
- default.qubit.jax (PennyLane-0.28.0)
- default.qubit.tf (PennyLane-0.28.0)
- default.qubit.torch (PennyLane-0.28.0)
- default.qutrit (PennyLane-0.28.0)
- null.qubit (PennyLane-0.28.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
Bug #3741 explains the background, but essentially qml.probs() does not order wires correctly. The desired wire order was given to an array, essentially saying "rearrange your axes to look like this", but that's not exactly how array indexing works. The arbitrary rearrangement of axes should instead be done with the `transpose(array, axes)`.

**Description of the Change:**
- Fix the wire rearranging logic in `qml.probs()` by using `np.transpose()`
- Update `QubitStateVector.state_vector()` to use this same logic.
  - I can do this in another PR but it feels related enough here. Context: I wrote that code a couple weeks ago and I thought to myself "hmmm why can't I rearrange my current axes to some arbitrary target axes? roll and swap just don't cut it! I'll just write my own basic axis-shuffler." Because the root of this bug is the same axis challenge, I figured I should update QSV as well

**Benefits:**
- `qml.probs(wires=new_order)` behaves as expected
- the code in `QubitStateVector.state_vector()` is prettier 😄 

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
Fixes #3741

### CODE REVIEW

File: pennylane/measurements/probs.py [hunk 1]
```diff
@@ -299,23 +300,24 @@ def marginal_prob(self, prob, wire_order, batch_size):
         # reshape the probability so that each axis corresponds to a wire
         num_device_wires = len(wire_order)
         shape = [2] * num_device_wires
+        desired_axes = np.argsort(np.argsort(mapped_wires))
```
Comment (timmysilv) [inline pennylane/measurements/probs.py:303]: using `np` instead of `qml.math` here because some interfaces don't have `np.insert` which I'm using below when batching

**Comment by timmysilv** [comment]
fwiw, I just found out that `pennylane_cirq.SimulatorDevice._expand_state` [expanded and re-arranged nicely](https://github.com/PennyLaneAI/pennylane-cirq/blob/4c09d20cb4c386d6912aa1d8c58253747accd0a7/pennylane_cirq/simulator_device.py#L117-L132) (but specifically for state-prep ops). `tranpose` is more concise, but still thought this was cool to find (I am deleting it in another PR though : D )

**Comment by timmysilv** [comment]
Tbh I wouldn't unless it's too burdensome. I could do something in this PR, but I hesitated because the qubit/qutrit devices should disappear and be replaced with the new Device API, and they will use the `qml.probs` implementation. That leaves only `probs.process_state` and `QSV.state_vector`, and ultimately the logic is just `np.transpose` and that's already bundled all nice for us

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -1670,35 +1609,29 @@ def marginal_prob(self, prob, wires=None):
...
-        powers_of_two = 2 ** np.arange(len(device_wires))[::-1]
-        perm = basis_states @ powers_of_two
-        # The permutation happens on the last axis both with and without broadcasting
-        return self._gather(prob, perm, axis=1 if batch_size is not None else 0)
+        # prob now is reshaped to have self.num_wires+1 axes in the case of broadcasting
+        prob = self._reshape(prob, shape)
+        # sum over all inactive wires
+        prob = self._reduce_sum(prob, inactive_device_wires)
+        # rearrange wires to desired order
+        prob = self._transpose(prob, desired_axes)
```
Comment (rmoyard) [inline pennylane/_qubit_device.py:1632]: Good finding, how efficient is this?

File: pennylane/measurements/probs.py [hunk 3]
```diff
@@ -299,23 +300,24 @@ def marginal_prob(self, prob, wire_order, batch_size):
         # reshape the probability so that each axis corresponds to a wire
```
Comment (rmoyard) [inline pennylane/measurements/probs.py:300]: Can you remind me, why do we have `marginal_probs` in the device and in the measurement? Which one is used when?

File: pennylane/measurements/probs.py (hunk 1 above)
Comment (rmoyard) [reply pennylane/measurements/probs.py:303]: Yes I think it should not be an issue because it is only to compute the axes and not the actual array and wires should not be specific to an interface.

File: pennylane/measurements/probs.py [hunk 4]
```diff
@@ -299,23 +300,24 @@ def marginal_prob(self, prob, wire_order, batch_size):
...
+            flat_shape = (batch_size, -1)
+
         # prob now is reshaped to have self.num_wires+1 axes in the case of broadcasting
         prob = qml.math.reshape(prob, shape)
 
         # sum over all inactive wires
-        flat_shape = (-1,) if batch_size is None else (batch_size, -1)
-        prob = qml.math.reshape(qml.math.sum(prob, axis=tuple(inactive_wires)), flat_shape)
+        if inactive_wires:
+            prob = qml.math.sum(prob, axis=tuple(inactive_wires))
```
Comment (rmoyard) [inline pennylane/measurements/probs.py:316]: What is the difference with the `reduced_sum` used before?

**Review by rmoyard — COMMENTED** [review]
Great job @timmysilv :100: A couple of questions, happy to approve after that!

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (timmysilv) [reply pennylane/_qubit_device.py:1632]: just tried something basic, it seems to be notably slower than on master... but master is wrong :) I can try something new like what was being done by the Cirq plugin (linked above). FWIW, here's my basic profiling technique:
```python
state = np.array([1 / 2, 0] * 4)
dev_wires = qml.wires.Wires(range(3))

def profile_wires(state, dev_wires):
    for probs_wires in ((0,1), (1, 2), (0, 2), (2, 0), (2, 1), (1, 2, 0), (2, 1, 0)):
        _ = qml.probs(wires=probs_wires).process_state(state, dev_wires)

%timeit profile_wires(state, dev_wires)
```

Master: `416 µs ± 5.22 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)`
Branch: `607 µs ± 9.81 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)`

File: pennylane/measurements/probs.py (hunk 3 above)
Comment (timmysilv) [reply pennylane/measurements/probs.py:300]: the MeasurementProcess `process_state` (and subsequently, `marginal_prob`) don't seem to be used at all today. It's effectively a copy of code, in anticipation for the device-specific code to be removed in the future. Today, the device entrypoint is [in staticstics()](https://github.com/PennyLaneAI/pennylane/blob/1774fe36f31f7de6266cf44983c85dfb561ec0fc/pennylane/_qubit_device.py#L818), which eventually finds its way to `device.marginal_prob`. In the future, we won't call `self.probability`. Instead, we'll call `probs_MP.process_state()`. But for today, I need to update both.

File: pennylane/measurements/probs.py (hunk 4 above)
Comment (timmysilv) [reply pennylane/measurements/probs.py:316]: this was actually just using `qml.math.sum` before as well, but yeah the old devices used `reduce_sum`. As you can see on the _qubit_device [here](https://github.com/PennyLaneAI/pennylane/blob/1774fe36f31f7de6266cf44983c85dfb561ec0fc/pennylane/_qubit_device.py#L113), it was just `np.sum` with axis provided. So we do the same thing here!

File: pennylane/measurements/probs.py (hunk 3 above)
Comment (rmoyard) [reply pennylane/measurements/probs.py:300]: Thanks, all good then :+1:

File: pennylane/_qubit_device.py (hunk 2 above)
Comment (rmoyard) [reply pennylane/_qubit_device.py:1632]: Feel free to explore :+1:  I do not think it is that bad, because prob should be called only once and now we have correct results.

Comment (timmysilv) [reply pennylane/_qubit_device.py:1632]: update! it wasn't transpose causing this difference, but instead the `if not np.all(desired_axes == range(len(desired_axes))):` check (I only did this check in `probs.py`, not for default.qubit). I removed the check in the latest commit, now the branch is more like 450us which is barely an overhead.

Comment (rmoyard) [reply pennylane/_qubit_device.py:1632]: Excellent @timmysilv :100:

**Review by rmoyard — APPROVED** [review]
Great bug fix :100:
